File: aws-lambda/activity_logger.py
import os
import json
import pymysql

def get_connection():
    """Establish and return a MySQL connection."""
    return pymysql.connect(
        host=os.environ.get('DB_HOST'),
        user=os.environ.get('DB_USER'),
        password=os.environ.get('DB_PASSWORD'),
        database=os.environ.get('DB_NAME'),
        cursorclass=pymysql.cursors.DictCursor
    )

def lambda_handler(event, context):
    """
    AWS Lambda Function: Activity Logger via API Gateway / SQS / Direct Invocation
    Connects to RDS MySQL to log events.
    Expected event structure (direct): { "user_id": 1, "action": "LOGIN", "details": "User logged in" }
    """
    connection = None
    try:
        connection = get_connection()
        logs_to_insert = []
        
        # Handle SQS batch events if triggered by SQS
        if 'Records' in event:
            for record in event['Records']:
                body = json.loads(record['body'])
                logs_to_insert.append((
                    body.get('user_id'), 
                    body.get('action'), 
                    body.get('details')
                ))
        else:
            # Direct invocation
            logs_to_insert.append((
                event.get('user_id'), 
                event.get('action'), 
                event.get('details')
            ))
            
        # Insert all logs into the database
        with connection.cursor() as cursor:
            sql = "INSERT INTO activity_logs (user_id, action, details) VALUES (%s, %s, %s)"
            cursor.executemany(sql, logs_to_insert)
            
        # Commit the transaction
        connection.commit()
        
        return {
            'statusCode': 200,
            'body': json.dumps({'message': 'Activity logged successfully'})
        }
        
    except Exception as e:
        print(f"Error logging activity: {e}")
        return {
            'statusCode': 500,
            'body': json.dumps({'error': str(e)})
        }
        
    finally:
        if connection:
            connection.close()

File: aws-lambda/activity-logger.js
// AWS Lambda Function: Activity Logger via API Gateway / SQS / Direct Invocation
// Connects to RDS MySQL to log events

const mysql = require('mysql2/promise');

exports.handler = async (event) => {
    // Expected event structure: { user_id, action, details }
    
    const dbConfig = {
        host: process.env.DB_HOST,
        user: process.env.DB_USER,
        password: process.env.DB_PASSWORD,
        database: process.env.DB_NAME,
    };
    
    let connection;
    try {
        connection = await mysql.createConnection(dbConfig);
        
        let logsToInsert = [];
        
        // Handle SQS batch events if triggered by SQS
        if (event.Records) {
            for (const record of event.Records) {
                const body = JSON.parse(record.body);
                logsToInsert.push([body.user_id, body.action, body.details]);
            }
        } else {
            // Direct invocation
            logsToInsert.push([event.user_id, event.action, event.details]);
        }
        
        for (const log of logsToInsert) {
            await connection.execute(
                'INSERT INTO activity_logs (user_id, action, details) VALUES (?, ?, ?)',
                log
            );
        }
        
        return {
            statusCode: 200,
            body: JSON.stringify({ message: 'Activity logged successfully' }),
        };
    } catch (error) {
        console.error('Error logging activity:', error);
        return {
            statusCode: 500,
            body: JSON.stringify({ error: error.message }),
        };
    } finally {
        if (connection) {
            await connection.end();
        }
    }
};